java/ql/lib/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -434,6 +436,71 @@ private module ControlFlowGraphImpl {
     )
   }
 
+  /**
+   * Gets the `i`th `SwitchCase` defined on `switch`, if one exists.
+   */
+  private SwitchCase getCase(StmtParent switch, int i) {
+    result = switch.(SwitchExpr).getCase(i) or result = switch.(SwitchStmt).getCase(i)
+  }
+
+  /**
+   * Gets the `i`th `PatternCase` defined on `switch`, if one exists.
+   */
+  private PatternCase getPatternCase(StmtParent switch, int i) {
+    result =
+      rank[i + 1](PatternCase pc, int caseIdx | pc = getCase(switch, caseIdx) | pc order by caseIdx)
+  }
+
+  /**
+   * Gets the PatternCase after pc, if one exists.
+   */
+  private PatternCase getNextPatternCase(PatternCase pc) {
+    exists(int idx, StmtParent switch |
+      pc = getPatternCase(switch, idx) and result = getPatternCase(switch, idx + 1)
+    )
+  }
+
+  /**
+   * Gets a `SwitchCase` that may be `pred`'s direct successor.
+   *
+   * This means any switch case that comes after `pred` up to the next pattern case, if any, except for `case null`.
+   *
+   * Because we know the switch block contains at least one pattern, we know by https://docs.oracle.com/javase/specs/jls/se21/html/jls-14.html#jls-14.11
+   * that any default case comes after the last pattern case.
+   */
+  private SwitchCase getASuccessorSwitchCase(PatternCase pred) {
+    result.getParent() = pred.getParent() and
```

Comment:
This is needlessly quadratic, but I guess we'll have to hit at least thousands of `PatternCase`s in a single switch before this starts to bite, so I'm ok leaving it as-is for now.

Reply:
Explain how? I thought by only computing it for PatternCase and having each result only appear once I would dodge that, but I guess you mean the intermediate result will be quadratic until the limit on the max index (that of the next pattern case) is applied?

Reply:
> the intermediate result will be quadratic until the limit on the max index (that of the next pattern case) is applied?

Yes. The constraints on index (both lower and upper bounds) can only be applied as filters once the product has been constructed. We would instead need to construct the index range and then do a two-column join on parent+index. E.g. something like this:
```
pragma[nomagic]
private predicate switchCaseIndex0(StmtParent switch, SwitchCase c, int i) {
  c.getParent() = switch and c.getIndex() = i
}

private int lastIndex(StmtParent switch) { result = max(int i | switchCaseIndex0(switch, _, i) }

pragma[inline_late]
bindingset[switch, i]
private predicate switchCaseIndex(StmtParent switch, SwitchCase c, int i) {
  switchCaseIndex0(switch, c, i)
}

private SwitchCase getASuccessorSwitchCase(PatternCase pred) {
  // Note we do include `case null, default` (as well as plain old `default`) here.
  not result.(ConstCase).getValue(_) instanceof NullLiteral and
  exists(int end |
    if exists(getNextPatternCase(pred))
    then end = getNextPatternCase(pred).getIndex()
    else end = lastIndex(pred.getParent())
  |
    switchCaseIndex(pred.getParent(), result, [pred.getIndex() + 1 .. end])
  )
}
```
Alternatively, use a step relation:
```
private SwitchCase nextCase(SwitchCase c) {
  exists(StmtParent switch, int i | c = getCase(switch, i) and result = getCase(switch, i + 1))
}

private SwitchCase nextCaseFromNonPatternCase(SwitchCase c) {
  result = nextCase(c) and not c instanceof PatternCase
}

private  SwitchCase getASuccessorSwitchCase(PatternCase pred) {
  result = nextCaseFromNonPatternCase*(nextCase(pred)) and
   // Note we do include `case null, default` (as well as plain old `default`) here.
  not result.(ConstCase).getValue(_) instanceof NullLiteral
}
```
The latter fix is simpler, but depends slightly on the optimiser to do something reasonable with the TC.

Reply:
The latter could also be expanded to an explicit recursion, which would guarantee the desired asymptotic behaviour, but then the null case becomes slightly awkward - unless we could possible push its exclusion into the rank in `getCase` - depends on the needs of the potential other uses of `getCase`, I haven't checked that.

Reply:
Gone with your first solution, only with `switchCaseIndex0` turned into a member `isNthCaseOf` in the spirit of `isNthChildOf`.